Reimplement navbar as per #246

VeeGee-Creations opened the pull request:
* switched from fixed position menu to flexbox layout. should now be responsive to more screen sizes.
* animation for hamburger seemed broken so I used on of my favorites. simply visual and can be changed as desired.
* removed the checkbox implementation of hamburger menu and replaced with state to apply active class.
* there were colors overwriting each other I used the ones tagged important but kept overwrites as comments. not sure which was intentional.
* fixed an issue with the entire nested ul highlighting at once.
* only one menu markup is used. on page break, css reshapes it to fit the modal.
* noticed on smaller screens that modal links don't fit height. added scrolling to try and address this issue.

I'm open to any changes or suggestions. I just slapped this together so I'm sure there are improvements to be made.

Closes #246 and #247.

VeeGee-Creations commented:
sorry about all the scss edits. the build process didn't catch the indentation or alphabetization

VeeGee-Creations commented:
* Changed from class to functional
* Using router to highlight active link
* Hamburger menu breaks on about page
  * Error caused by officers not having unique key. Currently key is set to officer.email but no such data exists.

Of note: when I switched from class to functional, I got some unhandled compile errors. The fix to this was to clear my build folder and re-compile. Apparently something in cache was causing an issue.

mattxwang reviewed (commented):
Not sure if you want us to review this yet as it's still marked as draft, but let me see if I can address some of the things you brought up.

> sorry about all the scss edits. the build process didn't catch the indentation or alphabetization

No worries at all! We'll squash merge this PR, so the commit history won't be preserved.

> Error caused by officers not having unique key. Currently key is set to officer.email but no such data exists.

Great point, this looks like a regression introduced a while back. I think the one-line bandaid fix is to key by `officer.title` instead (longer-term, I'm going to change the data structure) - do you want to put that in this PR? If not, I can submit a quick fix, and then you can rebase off of the new `main`.

> Of note: when I switched from class to functional, I got some unhandled compile errors. The fix to this was to clear my build folder and re-compile. Apparently something in cache was causing an issue.

Hm, that's really interesting and should not be happening; let me see if I can replicate, maybe we've misconfigured something.

---

Do you want any help with other portions of the navbar, as you've noted some areas that still need to be resolved? Glad to lend a hand, though my free time is getting vanishingly small :( 

P.S. - I noticed you've edited the Footer here as well; I'd move that change out of this PR, especially with #236 - it'll cause a merge conflict.

mattxwang commented:
> I appreciate your patience on this. I've reset head and merged. I'm not sure if it accomplished the goal as my rflog doesn't go back that far. It is going to require another force push though. Should I proceed with this? When I originally did the rebase I used `git rebase --onto`

Hm, I'm not super sure what's happening locally for you, but feel free to make a copy of your branch / relevant changes, then force-push to your fork; if something goes awry, we can always revert back to your branch copy, or cherry-pick your changes back into a fresh branch. Since you're on a fork, none of your changes are going to directly affect us upstream, so feel free to mess around as needed!

mattxwang commented:
Another strategy is that you can locally squash (using interactive rebase, or something else) your commits for the navbar into one commit. Then, we can cherry-pick that into a new branch off of the new `main`, which should work with no-frills!

VeeGee-Creations commented:
I wasn't comfortable enough with rebase to do a squash. I need to continue reading on the process. I did make a backup of the branch before pushing. Just when I was feeling pretty confident about using git I realized I had only scratched the surface!

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
-}
+export default function Navbar() {
+  // set states
+  const [menuActive, setMenuActive] = useState('');
```
Since this is functionally a boolean (it's either `"active"` or `""`), could we make this itself a boolean state variable, and then toggle the CSS class based off of the value of the boolean? I'm a bit iffy on "stringly-typed" variables, and would like to avoid them if possible!

i.e.

```js
const [isMenuActive, setIsMenuActive] = useState(false);
// ...
const toggleMenu = () => {
    setIsMenuActive(!isMenuActive);
};
// ...
const menuActiveClass = isMenuActive ? 'active' : '';

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+export default function Navbar() {
+  // set states
+  const [menuActive, setMenuActive] = useState('');
+  const [linkActive, setLinkActive] = useState('');
```
Nit, but could we rename this to `activeLink`/`setActiveLink`? I think it might fit more naturally.

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
-		}
-	}
+    setLinkActive(router.asPath);
+    hamburger.addEventListener('click', menuActivate);
```
Instead of adding this through the DOM (which I'd like to avoid if possible), could we instead bind this to the corresponding JSX element?

i.e.

```jsx
<div className={menuActive} id="hamburger" 
// ...
onClick={menuActivate} // or whatever you end up naming this as
>
```

That way, we can avoid manipulating the DOM outside of React, _and_ we don't need to write our own cleanup code!

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+    hamburger.addEventListener('keyup', menuAccessible);
+    // closes menu on link click
+    listItems.forEach((item) => {
+      item.addEventListener('click', menuActivate);
```
Same comment on moving the `'click'` event to the React `onClick` handler.

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+          </a>
+        </Link>
+        <section id="nav-items-container">
+          <div className={menuActive} id="hamburger" role="button" aria-label="navigation menu" aria-expanded={menuActive === 'active' ? true : false} tabIndex="0">
```
quick question - is there a reason that this is a `<div>` instead of a `<button>`? In general, it'd be great if we could use `<button>` instead, though I see you've set the `role` which helps with accessibility!

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+          <section className={menuActive} id="menu-modal">
+            <ul className={`nav-items ${menuActive}`} role="presentation">
+              <li><Link href="/about"><a className={linkActive === '/about' ? 'active' : ''}>About</a></Link></li>
+              <li><Link href="/committees"><a className={linkActive === '/commitees' ? 'active' : ''}>Committees</a></Link></li>
```
Typo:
```suggestion
              <li><Link href="/committees"><a className={linkActive === '/committees' ? 'active' : ''}>Committees</a></Link></li>
```

mattxwang reviewed (changes requested):
Just getting around to this now. Overall Vincent, really impressed, this looks great! Mostly only left some small code style/best-practices comments in the individual reviews.

Also, from a design aspect, I like the new animation! I'm a bit iffy on the circle - how does it look when you remove it (and just have the three bars, similar to what we had earlier)?

Going to tag in someone else from our dev team just to take a once-over on how the new navigation looks, and offer any other feedback. Overall, pretty happy with how this is!

mattxwang commented:
Oh, and on: 

> there were colors overwriting each other I used the ones tagged important but kept overwrites as comments. not sure which was intentional.

I think you've gotten the right ones. Quite a bit of the CSS is old/conflicting (and comes from before we had module-scoping as an option), so thank you for bearing with it!

mattxwang commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+        item.removeEventListener('keyup', menuAccessible);
+      });
+    };
+  }, [menuActive]);
```
Oh, one final note - not sure if `menuActive` is needed here as a dependency, since I imagine we only want to add/remove listeners on page load/exit, not when the actual menu gets opened or closed? But I might be misunderstanding!

VeeGee-Creations replied:
We're calling a page render everytime menu active updates in order to close the menu on click. I didn't originally have it there but noticed during testing that the menu didn't close.

VeeGee-Creations commented:
The circle is totally optional! It's just a simple border removal. I'll make the edit tonight

evanzhong commented on styles/components/Navbar.scss:
```diff
@@ -3,199 +3,207 @@
+        
+        &:hover {
+          // which color to keep?
+          // color: #999;
```
To keep the codebase clean, I'd delete this functional comment before merging
```suggestion
```

evanzhong commented on styles/components/Navbar.scss:
```diff
@@ -3,199 +3,207 @@
-    }
+        &:hover {
+          // prevents all committees from highlighting
+          color: $font-primary-color !important;
```
I think it would also be super cool if you could make each committee's hover color that committee's primary color (i.e. hack is purple, ai is light blue). Might be out of scope of this PR, but a good followup issue if it's not done here!

evanzhong reviewed (changes requested):
Great work so far, really appreciate the work you've done @VeeGee-Creations!

The majority of my review is left in comments. @mattxwang and I overlap of several items, which should be quick fixes. 

Overall items are
- Some code cleanup, levering React's functionality so you don't have to write a bunch of your own handlers and cleanup functions
- Abstracting away the logic you use to apply the `active` class to committee `<Link>`s
- Possibly Componentizing the hamburger menu UI element, or at least isolating the CSS tricks we use to a separate object or file. Has additional plus of reducing our reliance on `!important`
- Other minor UI suggestions

mattxwang commented on styles/components/Navbar.scss:
```diff
@@ -3,199 +3,207 @@
+          transform: translateY(8px);
+        }
+
+        &::after {
```
I'm personally ambivalent, is there any self-encapsulated state logic?

evanzhong replied:
No state logic for the hamburger stuff, some minor functional logic for color changes on hover.

I'm okay with not componentizing this, but do want to make this scss file more readable if someone isn't familiar with the `::before` and `::after` selectors

VeeGee-Creations replied:
With the ```::before``` and ```::after``` selectors. Are you wanting them to be written out as ```.bar::before``` or scrapped in favor of creating more divs in the jsx?

mattxwang commented:
Sounds good Vincent, and the changes look good so far - take your time!

VeeGee-Creations commented on components/Navbar.js:
```diff
@@ -1,109 +1,90 @@
+                  <li><Link href={'/committees#teachla'}><a className={linkActive === '/committees#teachla' ? 'active' : ''}>Teach LA</a></Link></li>
+                  <li><Link href={'/committees#w'}><a className={linkActive === '/committees#w' ? 'active' : ''}>W</a></Link></li>
+                  <li><Link href={'/committees#ai'}><a className={linkActive === '/committees#ai' ? 'active' : ''}>AI</a></Link></li>
+                  <li><Link href={'/committees#hack'}><a className={linkActive === '/committees#hack' ? 'active' : ''}>Hack</a></Link></li>
```
As far as I can see, there are 3 ways we can change the active state of these links. I'm fine with any of them so I'm going to leave this up to you.

Option 1:
    We can use javascript to manipulate the dom, adding and removing the class.

Option 2:
    We can go with inline logic for a more static approach.

Option 3:
    We can have a wrapper component around the each link which will apply the logic based on props passed from each link.

I know we want to minimize dom manipulation so I would say that option 3 is our best way to avoid that while bringing the logic out of inline.

evanzhong replied:
Let's got with option 3, you make a good argument in its favor

VeeGee-Creations commented:
Removed commented CSS

Committees hover and active color matches committee color

I can't replicate the first load transition issue with the hamburger on Chrome, Firefox or Edge. I have removed the transition and only left it on the bars. Hopefully this resolves it.

Removed background color transition on middle bar. Effect was not working as intended

This should be functional as is. I still need to make edits that resolve the implementation of hamburger bars and active link logic. I've commented those change request looking for more details on how you'd like them handled.

evanzhong commented on styles/components/Navbar.scss:
```diff
@@ -3,199 +3,240 @@
+          color: #ED3266;
+        }
+
+        [href='/committees#teachLA']:hover,
+        [href='/committees#teachLA'].active {
```
Small bug here due to a typo. 

```suggestion
        [href='/committees#teachla']:hover,
        [href='/committees#teachla'].active {
```

evanzhong reviewed (commented):
> With the ::before and ::after selectors. Are you wanting them to be written out as .bar::before or scrapped in favor of creating more divs in the jsx?
I'm okay with either option, if you're willing to create move divs in the jsx, I have a slight preference for that.

---

Noticing a small bug (or possibly a feature?) on mobile with the `active` class for the top right hamburger menu
Is the intended behavior to have it stay highlighted blue after the user opens the menu?
![image](https://user-images.githubusercontent.com/23728476/132102812-33bd1b91-72c8-4c31-8cb6-2bed8e28a31f.png)
![image](https://user-images.githubusercontent.com/23728476/132102828-0c7a7d33-4d2f-4188-951e-5042d0e15f9a.png)

Otherwise looking good!

mattxwang commented on styles/components/Navbar.scss:
```diff
@@ -3,199 +3,240 @@
+          padding: 0;
+        }
+
+        // committee color applied on hover and active
```
Two quick things here:

* this is quite a bit of duplication! I'm fine with this (since it's hard), but you may want to look into [looping with `@for` in SASS](https://sass-lang.com/documentation/at-rules/control/for). we don't have to do this if it's confusing to implement though
* instead of using these hex colors, could we use the [SASS variables for each committee](https://github.com/uclaacm/website/blob/main/styles/variables.scss)?

mattxwang reviewed (commented):
Wrapping up really well! Left one small code comment; also, I noticed that the fonts for the links in the nav are now off, as they're being overridden by the `button` default styling by a browser; mine are in Arial:

<img width="635" alt="Screen Shot 2021-09-04 at 2 06 56 PM" src="https://user-images.githubusercontent.com/14893287/132108042-daefb0b7-c36b-4435-9941-2c0092e0269d.png">
<img width="605" alt="Screen Shot 2021-09-04 at 2 07 06 PM" src="https://user-images.githubusercontent.com/14893287/132108043-c2f26dfc-3d19-40f0-bd3e-bfb6deba330f.png">

We could either switch off of `button` (though as I understand it, there's a reason you swapped to it), or we can use a more specific style to force the correct font inheritance. Open to either approach!

VeeGee-Creations commented:
inherited button font-family

committee colors use variables

committee and color added to list and iterated through with @each to apply styling

changed hamburger bars from ```::before``` and ```::after``` to jsx

created ActiveLink component which can be used in place of Link for links that need active styling
 * can add PropTypes to ensure activeClassName is set but will have to import prop-types as a dependency. I thought I should check before proceeding

i see the hamburger staying highlighted when simulating phone taps but not when using a mouse. I'll set up my network to let me access the page and test it on a real mobile device.

mattxwang commented:
From a behaviour perspective, this is looking amazing! I think the only thing that's off is that in the desktop view, the button for member login should have `font-weight: bold;`; other than that, it matches feature parity + is significantly better than what we've had previously. Great work Vincent!

> can add PropTypes to ensure activeClassName is set but will have to import prop-types as a dependency. I thought I should check before proceeding

Hm, I would prefer to not add prop-types at the moment; while types are a huge headache for us, I think what will eventually happen is that we'll switch to TS. In the meantime, I'm fine with leaving it as-is (misuse isn't necessarily breaking), though I might have a code review comment that neatly resolves this. 

> I'll set up my network to let me access the page and test it on a real mobile device.

You can use the [Netlify deploy](https://deploy-preview-253--jovial-pasteur-581b4a.netlify.app/) to preview your changes without using network access!

mattxwang commented:
Hey @VeeGee-Creations, just wanted to check in - how are you doing / do you need any help? Looking to get this in sometime soon, though there's no stress if you're unable to - we can wrangle some stuff together.

VeeGee-Creations commented:
Hey sorry, they switched me from nights to days at work and it took some time to adjust. I will try to get some more done this week but definitely by this weekend.

mattxwang commented:
> Hey sorry, they switched me from nights to days at work and it took some time to adjust. I will try to get some more done this week but definitely by this weekend.

Gotcha, sounds good. I'll sneak in a quick code review before then, but no worries if you're not able to wrap this up on the weekend; I can take over some of it if needed!

VeeGee-Creations commented:
I think I've finished the last of the changes needed. The members login link now displays proper casing and weight. After messing with the hamburger menu a bit, I've removed the active highlight. It still has slightly off behavior on mobile do to the nature of touch but will remove it's highlight when the user touches anywhere else on the screen (even scrolling). Highlighting functions as expected with mouse. If there is any other changes needed please let me know. I'm off Monday and will be able to apply them.

mattxwang reviewed (approved):
Vincent, I think you've done a stellar job! Thank you for working on this during your free time, it is much appreciated.

I have some maybe minor nits, but overall I have no large pieces of critical feedback. I think this is good to merge!

(to quickly write out the minor nits - am interested to talk about the design of `ActiveLink`, whether or not the links should be `<a>` or `<button>`, and some minor CSS things)

If @evanzhong is free, he can wrap up with any other comments. If not, let's squash and merge this in. Thank you again for all your hard work on this Vincent!

mattxwang commented:
Looks like no review from Evan, so I'm going to go ahead and merge. Again, great job @VeeGee-Creations, really appreciate it!

mattxwang commented:
@tnawathe21 can you rebase your navbar changes for "initiatives" off of this, and then open a PR for it?

VeeGee-Creations commented:
No problem! I'm going through some training at work but once everything settles back down I'll tackle some more issues
